core/src/main/scala/org/apache/spark/deploy/SparkS3Util.scala

```diff
@@ -0,0 +1,293 @@
+    if (key.startsWith("/")) {
+      key = key.substring(1)
+    }
+    if (key.endsWith("/")) {
+      key = key.substring(0, key.length - 1)
+    }
+    key
+  }
+
+  private def listPrefix(s3: AmazonS3Client, path: Path): Iterator[LocatedFileStatus] = {
+    val uri: URI = path.toUri
+    val key: String = keyFromPath(path)
+    val request: ListObjectsRequest = new ListObjectsRequest()
+      .withBucketName(uri.getAuthority)
+      .withPrefix(key)
+
+    val listings = new AbstractSequentialIterator[ObjectListing](s3.listObjects(request)) {
+      protected def computeNext(previous: ObjectListing): ObjectListing = {
+        if (!previous.isTruncated) {
+          return null
+        }
+        s3.listNextBatchOfObjects(previous)
+      }
+    }.asScala
+
+    val bucket: String = uri.getScheme + "://" + uri.getAuthority
+    listings
+      .map(listing => statusFromObjects(bucket, listing.getObjectSummaries))
+      .reduceLeft(_ ++ _)
+  }
+
+  private def statusFromObjects(
+      bucket: String,
+      objects: util.List[S3ObjectSummary]): Iterator[LocatedFileStatus] = {
+    val blockSize: Int = conf.getInt("fs.s3n.block.size", 67108864)
+    val list: ArrayBuffer[LocatedFileStatus] = ArrayBuffer()
+    for (obj: S3ObjectSummary <- objects.asScala) {
+      if (!obj.getKey.endsWith("/")) {
+        val status: FileStatus = new FileStatus(
+          obj.getSize, false, 1, blockSize,
+          obj.getLastModified.getTime,
+          new Path(bucket + "/" + obj.getKey))
+        list += createLocatedFileStatus(status)
+      }
+    }
+    list.iterator
+  }
+
+  private def createLocatedFileStatus(status: FileStatus): LocatedFileStatus = {
+    val fakeLocation: Array[BlockLocation] = Array[BlockLocation]()
+    new LocatedFileStatus(status, fakeLocation)
+  }
+
+  private def listLocatedStatus(
+      s3: AmazonS3Client,
+      path: Path): RemoteIterator[LocatedFileStatus] = {
+    new RemoteIterator[LocatedFileStatus]() {
+      private final val iterator: Iterator[LocatedFileStatus] = listPrefix(s3, path)
+
+      def hasNext: Boolean = {
+        try {
+          iterator.hasNext
+        } catch {
+          case e: AmazonClientException => throw new IOException(e)
+        }
+      }
+
+      def next(): LocatedFileStatus = {
+        try {
+          iterator.next()
+        } catch {
+          case e: AmazonClientException => throw new IOException(e)
+        }
+      }
+    }
+  }
+
+  /**
+   * List paths in parallel and merge returned `FileStatus` objects into a single list.
+   */
+  private def listStatus(
+      s3: AmazonS3Client,
+      paths: List[String],
+      parallelism: Int): Array[FileStatus] = {
+    val list: ArrayBuffer[FileStatus] = ArrayBuffer()
+    val parArray: ParArray[String] = paths.toParArray
+    parArray.tasksupport = new ForkJoinTaskSupport(new ForkJoinPool(parallelism))
+    parArray
+      .map { path => listLocatedStatus(s3, new Path(path)) }
```

Comment:
Is the AmazonS3Client thread safe? Or should we use a client for each thread (mapper)?